### PR TITLE
Start with conversations

### PR DESCRIPTION
**Pull request by rodrieiz**
https://trello.com/c/PYcxvv4Z/10-3-as-a-user-i-should-be-able-to-see-all-my-conversations

### CODE REVIEW

File: app/controllers/conversations_controller.rb [hunk 1]
```diff
@@ -0,0 +1,7 @@
+class ConversationsController < ApplicationController
+  before_action :authenticate_user!
```
Comment (santir489) [inline app/controllers/conversations_controller.rb]: You can move this to `ApplicationController` WDYT?

File: spec/models/conversation_spec.rb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe Conversation, type: :model do
+  pending "add some examples to (or delete) #{__FILE__}"
```
Comment (santir489) [inline spec/models/conversation_spec.rb]: What about some model validations?

File: spec/requests/conversations_spec.rb [hunk 3]
```diff
@@ -0,0 +1,42 @@
+require 'rails_helper'
+
+RSpec.describe 'GET conversation/index', type: :request do
+  let!(:user) { create(:user) }
+  let!(:user2) { create(:user) }
+  let!(:topic) { create(:topic) }
+  # let!(:target1) { create(:target, topic_id: topic.id, user: user, latitude: 90, longitude: 90) }
```
Comment (santir489) [inline spec/requests/conversations_spec.rb]: Please do not commit commented code

File: spec/requests/conversations_spec.rb [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  let!(:target2) { create(:target, topic_id: topic.id, user: user2, latitude: 90, longitude: 90) }
+  let!(:target3) { create(:target, topic_id: topic.id, user: user2, latitude: 5, longitude: 5) }
+
+  subject(:create_target) { post targets_url, params: params, headers: user_auth_headers, as: :json }
```
Comment (santir489) [inline spec/requests/conversations_spec.rb]: If conversation creation happens when a target is created, does it makes sense to test it on `request/create_target_spec` ?

Comment (rodrieiz) [reply spec/requests/conversations_spec.rb]: Could be but I added a match here to test the get conversations include it in the response

File: app/controllers/targets_controller.rb [hunk 5]
```diff
@@ -1,9 +1,13 @@
 class TargetsController < ApplicationController
-  before_action :authenticate_user!
-
   def create
     @target = current_user.targets.create(target_params)
     if @target.save
+      @possible_matches = Target.within(@target.radius, origin: @target)
+      @possible_matches = @possible_matches.where(@target.topic == :topic)
+      @possible_matches.each do |item|
+        Conversation.create(user1: User.find(item.user.id), user2: current_user) if item.user_id != current_user.id
```
Comment (DJA89) [inline app/controllers/targets_controller.rb]: Maybe you can `@possible_matches = @possible_matches.where(@target.topic == :topic).where.not(user: current_user)` and avoid the `if` below.

File: spec/models/conversation_spec.rb [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+#
+#  id         :bigint           not null, primary key
+#  user1      :integer          not null
+#  user2      :integer          not null
+#  created_at :datetime         not null
+#  updated_at :datetime         not null
+#
+require 'rails_helper'
+
+RSpec.describe Conversation, type: :model do
```
Comment (DJA89) [inline spec/models/conversation_spec.rb:13]: Y ou can try creating a conversation where both users are the same and check that it doesn't save and that it returns a correct error

File: spec/requests/conversations_spec.rb (hunk 4 above)
Comment (santir489) [reply spec/requests/conversations_spec.rb]: `I added a match here to test the get conversations ` you should do that using factories

File: app/models/conversation.rb [hunk 7]
```diff
@@ -0,0 +1,16 @@
+# == Schema Information
+#
+# Table name: conversations
+#
+#  id         :bigint           not null, primary key
+#  user1      :integer          not null
+#  user2      :integer          not null
+#  created_at :datetime         not null
+#  updated_at :datetime         not null
+#
```
Comment (sebastiancaraballo) [inline app/models/conversation.rb:10]: Try adding an index `user1, user2`. Take a look at [this](https://guides.rubyonrails.org/active_record_migrations.html).

File: app/controllers/targets_controller.rb [hunk 8]
```diff
@@ -1,9 +1,13 @@
 class TargetsController < ApplicationController
-  before_action :authenticate_user!
-
   def create
     @target = current_user.targets.create(target_params)
     if @target.save
+      @possible_matches = Target.within(@target.radius, origin: @target)
```
Comment (sebastiancaraballo) [inline app/controllers/targets_controller.rb]: This controller action is too big. Consider extracting this logic to a service.

File: app/models/target.rb [hunk 9]
```diff
@@ -16,5 +16,11 @@ class Target < ApplicationRecord
   belongs_to :topic
   belongs_to :user
 
+  acts_as_mappable defaul_units: :kms,
```
Comment (sebastiancaraballo) [inline app/models/target.rb]: typo

File: app/services/find_match.rb [hunk 10]
```diff
@@ -0,0 +1,15 @@
+class FindMatch
+  def initialize(target, user)
+    @target = target
+    @user = user
+  end
+
+  def search
+    @possible_matches = Target.within(@target.radius, origin: @target)
+    @possible_matches = @possible_matches.where(topic: @target.topic).where.not(user: @user)
+    @possible_matches.each do |item|
```
Comment (sebastiancaraballo) [inline app/services/find_match.rb]:
```suggestion
    @possible_matches.each do |target|
```

File: app/services/find_match.rb [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+      Conversation.create(user1: User.find(item.user.id), user2: @user)
```
Comment (sebastiancaraballo) [inline app/services/find_match.rb]:
```suggestion
      Conversation.create(user1: User.find(target.user_id), user2: @user)
```

File: app/models/conversation.rb (hunk 7 above)
Comment (sebastiancaraballo) [reply app/models/conversation.rb:10]: Try adding a new [migration](https://apidock.com/rails/v2.2.1/ActiveRecord/ConnectionAdapters/SchemaStatements/add_index) to validate at DB level as well:
`add_index :conversations, [:user1, :user2], unique: true`

Comment (rodrieiz) [reply app/models/conversation.rb:10]: I think I did it!

**Review by sebastiancaraballo — APPROVED** [review]

**Review by DJA89 — APPROVED** [review]

**Pull request merged**